test_conformance/generic_address_space/atomic_tests.cpp

```diff
@@ -0,0 +1,231 @@
+    test_error(err, "clCreateProgramWithSource failed");
+
+    err = clBuildProgram(program, 1, &deviceID, "-cl-std=CL2.0", nullptr,
+                         nullptr);
+    test_error(err, "clBuildProgram failed");
+
+    clKernelWrapper kernel = clCreateKernel(program, "testKernel", &err);
+    test_error(err, "clCreateKernel failed");
+
+    size_t wgSize, retSize;
+    err = clGetKernelWorkGroupInfo(kernel, deviceID,
+                                   CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE,
+                                   sizeof(wgSize), &wgSize, &retSize);
+    test_error(err, "clGetKernelWorkGroupInfo failed");
+
+    // How many workgroups to run for the test.
+    const int numWGs = 2;
+    const size_t bufferSize = numWGs * sizeof(cl_uint);
+    clMemWrapper buffer =
+        clCreateBuffer(context, CL_MEM_WRITE_ONLY, bufferSize, nullptr, &err);
+    test_error(err, "clCreateBuffer failed");
+    const cl_int zero = 0;
+    err = clEnqueueFillBuffer(queue, buffer, &zero, sizeof(zero), 0, bufferSize,
+                              0, nullptr, nullptr);
+    test_error(err, "clEnqueueFillBuffer failed");
+
+    err = clSetKernelArg(kernel, 0, sizeof(cl_mem), &buffer);
+    test_error(err, "clSetKernelArg failed");
+    err = clSetKernelArg(kernel, 1, bufferSize, nullptr);
+    test_error(err, "clSetKernelArg failed");
+
+    const size_t globalSize = wgSize * numWGs;
+    err = clEnqueueNDRangeKernel(queue, kernel, 1, nullptr, &globalSize,
+                                 &wgSize, 0, nullptr, nullptr);
+    test_error(err, "clEnqueueNDRangeKernel failed");
+
+    std::vector<cl_int> results(numWGs);
+    err = clEnqueueReadBuffer(queue, buffer, CL_TRUE, 0, bufferSize,
+                              results.data(), 0, nullptr, nullptr);
+    test_error(err, "clEnqueueReadBuffer failed");
+
+    clFinish(queue);
+
+    for (size_t i = 0; i < wgSize; ++i)
+    {
+        const cl_int expected = ((i % 2) == 0) ? wgSize - 1 : wgSize;
+        if (results[i] != expected)
+        {
+            log_error("Verification on device failed at index %zu\n", i);
+            return TEST_FAIL;
+        }
+    }
+
+    return CL_SUCCESS;
+}
+
+int test_generic_atomics_variant(cl_device_id deviceID, cl_context context,
+                                 cl_command_queue queue, int num_elements)
+{
+    const auto version = get_device_cl_version(deviceID);
+
+    if (version < Version(2, 0)) return TEST_SKIPPED_ITSELF;
+
+    cl_int err = CL_SUCCESS;
+    clProgramWrapper program = clCreateProgramWithSource(
+        context, 1, &KernelSourceVariant, nullptr, &err);
+    test_error(err, "clCreateProgramWithSource failed");
+
+    err = clBuildProgram(program, 1, &deviceID, "-cl-std=CL2.0", nullptr,
+                         nullptr);
+    test_error(err, "clBuildProgram failed");
+
+    clKernelWrapper kernel = clCreateKernel(program, "testKernel", &err);
+    test_error(err, "clCreateKernel failed");
```

Comment:
This should use `create_single_kernel_helper` too.

Reply:
Should be done now